pkg/apis/proto/source/v1/udsource.proto

```diff
@@ -0,0 +1,139 @@
+syntax = "proto3";
+
+option go_package = "github.com/numaproj/numaflow-go/pkg/apis/proto/source/v1";
+
+import "google/protobuf/timestamp.proto";
+import "google/protobuf/empty.proto";
+
+package source.v1;
+
+service UserDefinedSource {
+  // Read returns a stream of datum responses.
+  // The size of the returned ReadResponse is less than or equal to the num_records specified in ReadRequest.
+  rpc Read(ReadRequest) returns (stream ReadResponse);
+
+  // Ack acknowledges a list of datum offsets.
+  // It indicates that the datum stream has been processed by the source vertex.
+  rpc Ack(AckRequest) returns (AckResponse);
+
+  // Pending returns the number of pending records at the user defined source.
+  rpc Pending(google.protobuf.Empty) returns (PendingResponse);
+
+  // IsReady is the heartbeat endpoint for gRPC.
+  rpc IsReady(google.protobuf.Empty) returns (ReadyResponse);
```

Comment:
Let's name it IsSourceReady because eventually, we will also have Source Transformer, and it might need a different health check endpoint since both of them can be used independently or together.

Reply:
now that each will be a separate service, we can call it `IsReady` itself.